File: app/src/main/java/com/garyjacobs/weathertest/MainActivity.java
package com.garyjacobs.weathertest;

import android.Manifest;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.Toast;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationListener;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.squareup.otto.Subscribe;

import java.io.IOException;
import java.security.Permission;
import java.util.List;

import Events.ForecastListSelected;
import widgets.ComboBox;

public class MainActivity extends WeatherActivity implements GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener, LocationListener {

    public static int FORECASTDATA_FETCHED = 1;
    private GoogleApiClient mGoogleAPIClient;
    private LocationRequest locationRequest;
    FrameLayout weatherListContainer;
    FrameLayout weatherDetailsContainer;
    private ComboBox cityForecastCB;
    boolean twoPane;
    boolean isBound = false;
    IBinder iBinder;
    Messenger outBoundMessenger;
    Messenger inBoundMessenger;
    Message outBoundMessage;

    @Override
    protected void onStart() {
        super.onStart();
        // check if user has granted location permissions
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION}, 100);
        } else {
            mGoogleAPIClient.connect();
        }

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            mGoogleAPIClient.connect();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        mGoogleAPIClient.disconnect();
    }

    @Override
    protected void onResume() {
        super.onResume();
        getWeatherApplication().getBus().register(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        getWeatherApplication().getBus().unregister(this);
    }

    @Subscribe
    public void OnItemSelected(ForecastListSelected forecastListSelected) {
        if (twoPane) {
            LoadForecastDetailsFragment(forecastListSelected.getSelectedItem(), R.id.weather_details_container, false);
        } else {
            LoadForecastDetailsFragment(forecastListSelected.getSelectedItem(), R.id.weather_list_container, true);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mGoogleAPIClient = new GoogleApiClient.Builder(this)
                .addApi(LocationServices.API)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .build();

        // Create the LocationRequest object
        locationRequest = LocationRequest.create()
                .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
                .setInterval(10 * 1000)        // 10 seconds, in milliseconds
                .setFastestInterval(1 * 1000); // 1 second, in milliseconds


        FrameLayout fl = (FrameLayout) findViewById(R.id.main_frame_layout);
        cityForecastCB = (ComboBox) findViewById(R.id.location_cb);
        cityForecastCB.setClientClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //currentZip = ((TextView) v).getText().toString();
                if (isBound) {
                    Message message = Message.obtain();
                    message.replyTo = inBoundMessenger;
                    message.what = FetchForecastService.FETCH_CITY_FORECAST;
                    Bundle bundle = new Bundle();
                    bundle.putString("FETCH_CITY_FORECAST", cityForecastCB.getCurrentLocation());
                    message.setData(bundle);
                    try {
                        outBoundMessenger.send(message);
                    } catch (RemoteException re) {
                        Log.d(null, re.getMessage(), re);
                    }
                }

            }
        });
        View v = (View) getLayoutInflater().inflate(R.layout.weather_panel_layout, null);
        fl.addView(v);
        twoPane = (FrameLayout) v.findViewById(R.id.weather_details_container) != null;
    }

    private class IncomingHandler extends Handler {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case FetchForecastService.CLIENT_REGISTERED:
                    outBoundMessage = Message.obtain();
                    Bundle bundle = new Bundle();
                    bundle.putString("FETCH_CITY_FORECAST", cityForecastCB.getCurrentLocation());
                    outBoundMessage.setData(bundle);
                    outBoundMessage.what = FetchForecastService.FETCH_CITY_FORECAST;
                    try {
                        outBoundMessenger.send(outBoundMessage);
                    } catch (RemoteException e) {
                        e.printStackTrace();
                    }
                    break;
                case FetchForecastService.CITY_FORECAST_FETCHED:
                    LoadWeatherListFragment();
                    if (twoPane) {
                        LoadForecastDetailsFragment(0, R.id.weather_details_container, false);
                    }
                    getSupportActionBar().setTitle(getWeatherApplication().getForecast().getCity().getName());
                    break;
            }
        }
    }

    private void LoadWeatherListFragment() {
        FragmentManager fm = getFragmentManager();
        fm.beginTransaction().replace(R.id.weather_list_container, new WeatherListFragment()).commit();
    }

    private void LoadForecastDetailsFragment(int position, int containerID, boolean incCount) {
        FragmentManager fm = getFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(containerID, ForecastDetailsFragment.newInstance(position));
        if (incCount) {
            ft.addToBackStack("WeatherDetails");
        }
        ft.commit();
    }

    private ServiceConnection mServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            iBinder = service;
            isBound = true;
            outBoundMessenger = new Messenger(iBinder);
            inBoundMessenger = new Messenger(new IncomingHandler());
            Message msg = Message.obtain();
            msg.replyTo = inBoundMessenger;
            msg.what = FetchForecastService.REGISTER_CLIENT;
            try {
                outBoundMessenger.send(msg);
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            isBound = false;
        }
    };

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbindService(mServiceConnection);
    }

    @Override
    public void onBackPressed() {
        FragmentManager fm = getFragmentManager();
        int fragCount = fm.getBackStackEntryCount();
        if (fragCount > 0) {
            fm.popBackStack();
        } else {
            super.onBackPressed();
        }

    }

    @Override
    public void onConnected(Bundle connectionHint) {

        Location lastLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleAPIClient);
        if (lastLocation != null) {
            handleNewLocation(lastLocation);
        } else {
            LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleAPIClient, locationRequest, this);
        }

    }

    @Override
    public void onConnectionSuspended(int i) {

    }

    @Override
    public void onConnectionFailed(ConnectionResult connectionResult) {

    }

    @Override
    public void onLocationChanged(Location location) {
        handleNewLocation(location);
    }

    private void handleNewLocation(Location newLocation) {
        Geocoder geocoder = new Geocoder(this);
        try {
            List<Address> addressList = geocoder.getFromLocation(newLocation.getLatitude(), newLocation.getLongitude(), 1);
            cityForecastCB.setText(addressList.get(0).getLocality());
            if (!isBound) {
                // start ForcastFetcher service
                Intent intent = new Intent(this, FetchForecastService.class);
                bindService(intent, mServiceConnection, BIND_AUTO_CREATE);
            }
        } catch (IOException ioe) {
            Toast.makeText(this, "Error obtaining zipcode", Toast.LENGTH_LONG);
        }
    }
}

File: app/src/main/java/com/garyjacobs/weathertest/MainActivity.kt
package com.garyjacobs.weathertest

import android.Manifest
import android.app.FragmentManager
import android.app.FragmentTransaction
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.location.Address
import android.location.Geocoder
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Message
import android.os.Messenger
import android.os.RemoteException
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import android.widget.Toast

import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationListener
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.squareup.otto.Subscribe

import java.io.IOException

import Events.ForecastListSelected
import android.content.Context
import android.widget.ProgressBar
import widgets.ComboBox

class MainActivity : WeatherActivity(), GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener, LocationListener {
    private var mGoogleAPIClient: GoogleApiClient? = null
    private var locationRequest: LocationRequest? = null
    lateinit var weatherListContainer: FrameLayout
    lateinit var weatherDetailsContainer: FrameLayout
    lateinit var progress_bar: ProgressBar
    lateinit var cityForecastCB: ComboBox
    var twoPane: Boolean = false
    internal var isBound = false
    lateinit var iBinder: IBinder
    lateinit var outBoundMessenger: Messenger
    lateinit var inBoundMessenger: Messenger
    lateinit var outLocaterMessenger : Messenger
    lateinit var inLocaterMessenter: Messenger

    private val mServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            isBound = true
            outBoundMessenger = Messenger(service)
            inBoundMessenger = Messenger(IncomingHandler())
            cityForecastCB.visibility = View.VISIBLE
            outBoundMessenger.sendMessage(FetchForecastService.REGISTER_CLIENT)
        }

        override fun onServiceDisconnected(name: ComponentName) {
            isBound = false
        }
    }

    private val locaterServiceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(name: ComponentName?) {
        }

        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            outLocaterMessenger
        }
    }
    override fun onStart() {
        super.onStart()
        // check if user has granted location permissions
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED || ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION), 100)
        } else {
            mGoogleAPIClient!!.connect()
        }

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            mGoogleAPIClient!!.connect()
        }
    }

    override fun onStop() {
        super.onStop()
        //mGoogleAPIClient!!.disconnect()
    }

    override fun onResume() {
        super.onResume()
        weatherApplication.getBus().register(this)
    }

    override fun onPause() {
        super.onPause()
        weatherApplication.getBus().unregister(this)
    }

    @Subscribe
    fun OnItemSelected(forecastListSelected: ForecastListSelected) {
        if (twoPane) {
            LoadForecastDetailsFragment(forecastListSelected.selectedItem, R.id.weather_details_container, false)
        } else {
            LoadForecastDetailsFragment(forecastListSelected.selectedItem, R.id.weather_list_container, true)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mGoogleAPIClient = GoogleApiClient.Builder(this)
                .addApi(LocationServices.API)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .build()

        // Create the LocationRequest object
        locationRequest = LocationRequest.create()
                .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
                .setInterval((10 * 1000).toLong())        // 10 seconds, in milliseconds
                .setFastestInterval((1 * 1000).toLong()) // 1 second, in milliseconds
        progress_bar = findViewById(R.id.progress_bar) as ProgressBar
        val frameLayout = findViewById(R.id.main_frame_layout) as FrameLayout
        cityForecastCB = findViewById(R.id.location_cb) as ComboBox
        cityForecastCB?.setClientClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                if (isBound) {
                    cityForecastCB.getCurrentText()?.let {
                        val bundle = Bundle()
                        bundle.putString("FETCH_CITY_FORECAST", it)
                        outBoundMessenger.sendMessage(FetchForecastService.FETCH_CITY_FORECAST, bundle)
                    }
                }
            }
        })
        val v = layoutInflater.inflate(R.layout.weather_panel_layout, null) as View
        frameLayout.addView(v)
        twoPane = v.findViewById(R.id.weather_details_container) != null
    }


    inner class IncomingHandler : Handler() {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                FetchForecastService.CLIENT_REGISTERED -> {
                    cityForecastCB.getCurrentText()?.let {
                        val bundle = Bundle()
                        bundle.putString("FETCH_CITY_FORECAST", it)
                        outBoundMessenger.sendMessage(FetchForecastService.FETCH_CITY_FORECAST, bundle)
                    }
                }
                FetchForecastService.CITY_FORECAST_FETCHED -> {
                    progress_bar.visibility = View.INVISIBLE

                    LoadWeatherListFragment()
                    if (twoPane) {
                        LoadForecastDetailsFragment(0, R.id.weather_details_container, false)
                    }
                    supportActionBar!!.title = weatherApplication.getForecast().city.name
                }
            }
        }
    }


    fun Messenger.sendMessage(what: Int, bundle: Bundle? = null, outMessenger: Messenger = outBoundMessenger, inMessenger: Messenger = inBoundMessenger) {
        val outboundmessage = Message.obtain()
        outboundmessage.what = what
        outboundmessage.replyTo = inMessenger
        bundle?.let { outboundmessage.data = bundle }
        try {
            outMessenger.send(outboundmessage)
        } catch (e: RemoteException) {
            e.printStackTrace()
        }
    }

    private fun LoadWeatherListFragment() {
        val fm = fragmentManager
        fm.beginTransaction().replace(R.id.weather_list_container, WeatherListFragment()).commit()
    }

    private fun LoadForecastDetailsFragment(position: Int, containerID: Int, incCount: Boolean) {
        val fm = fragmentManager
        val ft = fm.beginTransaction()
        ft.replace(containerID, ForecastDetailsFragment.newInstance(position))
        if (incCount) {
            ft.addToBackStack("WeatherDetails")
        }
        ft.commit()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        val id = item.itemId


        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)

    }

    override fun onDestroy() {
        super.onDestroy()
        unbindService(mServiceConnection)
    }

    override fun onBackPressed() {
        val fm = fragmentManager
        val fragCount = fm.backStackEntryCount
        if (fragCount > 0) {
            fm.popBackStack()
        } else {
            super.onBackPressed()
        }

    }

    override fun onConnected(connectionHint: Bundle?) {

        val lastLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleAPIClient)
        if (lastLocation != null) {
            handleNewLocation(lastLocation)
        } else {
            LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleAPIClient, locationRequest, this)
        }

    }

    override fun onConnectionSuspended(i: Int) {
        Toast.makeText(this, "onConnectionSuspended val: $i", Toast.LENGTH_LONG)
    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {
        Toast.makeText(this, "onConnectionSuspended val: ${connectionResult.toString()}", Toast.LENGTH_LONG)
    }

    override fun onLocationChanged(location: Location) {
        handleNewLocation(location)
    }

    private fun handleNewLocation(newLocation: Location) {
        val geocoder = Geocoder(this)
        try {
            val addressList = geocoder.getFromLocation(newLocation.latitude, newLocation.longitude, 1)
            val address = addressList[0]
            cityForecastCB!!.setText(address.postalCode)
            if (!isBound) {
                // start ForcastFetcher service
                val intent = Intent(this, FetchForecastService::class.java)
                bindService(intent, mServiceConnection, Context.BIND_AUTO_CREATE)
            }
        } catch (ioe: IOException) {
            Toast.makeText(this, "Error obtaining zipcode", Toast.LENGTH_LONG)
        }

    }

    companion object {

        var FORECASTDATA_FETCHED = 1
    }
}

